js/connection.js

```diff
@@ -31,31 +34,143 @@ weechat.factory('connection',
+                if (!settings.allowPlaintextAuthentication)
+                {
+                    $rootScope.oldWeechatError = true;
+                    $rootScope.$emit('relayDisconnect');
+                    $rootScope.$digest() // Have to do this otherwise change detection doesn't see the error.
+                    throw new Error('Plainttext authentication not allowed.');
+                }
+
                 // Escape comma in password (#937)
                 passwd = passwd.replace(',', '\\,');
-                // This is not the proper way to do this.
-                // WeeChat does not send a confirmation for the init.
-                // Until it does, We need to "assume" that formatInit
-                // will be received before formatInfo
+
                 ngWebsockets.send(
-                    weeChat.Protocol.formatInit({
+                    weeChat.Protocol.formatInitPre29({
                         password: passwd,
                         compression: noCompression ? 'off' : 'zlib',
-                        useTotp: useTotp,
                         totp: totp
                     })
                 );
 
-                return ngWebsockets.send(
-                    weeChat.Protocol.formatInfo({
-                        name: 'version'
+                // Wait a little bit until the init is sent
+                return new Promise(function(resolve) {
+                    setTimeout(() => resolve(), 5);
+                })
+
+            };
+
+            // Helper methods for initialization commands
+            // This method is used to initialize weechat >= 2.9
+            var salt;
+            var _initializeConnection29 = function(passwd, nonce, iterations, totp) {
+
+                return window.crypto.subtle.importKey(
+
+                    'raw',
+                    utils.stringToUTF8Array(passwd),
+                    {name: 'PBKDF2'},//{name: 'HMAC', hash: 'SHA-512'},
+                    false,
+                    ['deriveBits']
+
+                ).then( function (key) {
+
+                    salt = utils.concatenateTypedArray(utils.concatenateTypedArray(nonce, new Uint8Array([0x3A])), window.crypto.getRandomValues(new Uint8Array(16))); //nonce:cnonce, 3A is a ':' in ASCII
+                    return window.crypto.subtle.deriveBits(
+                        {
+                            name: 'PBKDF2',
+                            hash: 'SHA-512',
+                            salt: salt,
+                            iterations: iterations,
+                        },
+                        key, //your key from generateKey or importKey
+                        512
+                    );
+                    
+                }).then( function (hash) {
+
+
+                    ngWebsockets.send(
+                        weeChat.Protocol.formatInit29(
+                            'pbkdf2+sha512:' + utils.bytetoHexString(salt) + ':100000:' + utils.bytetoHexString(hash),
```

Comment:
The iterations probably shouldn't be hardcoded here

Reply:
Yes, sorry, fixed now. Tested to see if it still works and it does.